### DISCUSSION TITLE
How to get query parameters in nuxt plugin ?

### DISCUSSION THREAD

**Discussion by johnymanuelli** [nuxt/nuxt#17542] [Questions]
Hey there, im trying to get query parameters in my custom nuxt plugin, but without success. 

FILE : @/plugins/cb-decrypt/index.js

```
import { defineNuxtPlugin} from '#app'
import { useRoute } from 'vue-router'

export default defineNuxtPlugin( (nuxtApp) => {
    const route = useRoute();


    console.log('CB_DECRYPT Plugin :: Mounted');

    // Here i cannot find query, route or something like that
    console.log('NUXT_APP :: ', nuxtApp.vueApp);

    // THIS Returns undefined
    console.log('useRoute :: ', route);

   .... more hidden code
});
```
Can someone help ?  Thank you.

**Comment by danielroe** [comment]
You can use the `useRoute` provided by Nuxt itself:
```js
import { defineNuxtPlugin } from '#app'
import { useRoute } from '#imports'

export default defineNuxtPlugin((nuxtApp) => {
  const route = useRoute()
  // this should now work as you wish
})
```

**Reply by johnymanuelli** [reply]
This solution still returns undefined : 

console.log('useRoute :: ', route);

<img width="952" alt="Snímek obrazovky 2021-12-20 v 13 13 31" src="https://user-images.githubusercontent.com/19531029/146765658-2f89d1c1-534b-40f5-9d09-0fb67dda5bd3.png">

**Reply by danielroe** [reply]
Just ensure you're using the latest version of nuxt, as we've only recently implemented a wrapped `useRoute` that works within plugins.

**Reply by johnymanuelli** [reply]
> Just ensure you're using the latest version of nuxt, as we've only recently implemented a wrapped `useRoute` that works within plugins.

Thank you, I've updated to the last version and now it's working. 👍 💯

**Comment by johnymanuelli** [comment]
@danielroe 
Hey there, I'm still fighting to get the query params from the URL. I've updated to the latest version of nuxt, this fixed the problem with useRoute. The useRoute it's working. 

But if i call 
`console.log('useRoute query :: ', route.query);`

Im not able to get the query params and get something like this.

<img width="511" alt="Snímek obrazovky 2021-12-20 v 15 35 46" src="https://user-images.githubusercontent.com/19531029/146783880-7363cd57-a698-4a52-b53c-ca427e1bb0db.png">

The same happens with:

`console.log('globalProperties.$route :: ', nuxtApp.vueApp.config.globalProperties.$route);`

<img width="775" alt="Snímek obrazovky 2021-12-20 v 15 56 33" src="https://user-images.githubusercontent.com/19531029/146787036-d0a4242c-8818-4f77-9d6a-01ee0760bdc0.png">


Any idea ? Thanks

**Comment by joemonkeylee** [comment]
Hi,there. I'd got the same question. Nuxt3 is the @latest version now. Here goes my enviroment info

------------------------------
- Operating System: `Darwin`
- Node Version:     `v14.17.6`
- Nuxt Version:     `3.0.0-27415326.3c563fa`
- Package Manager:  `yarn@1.22.17`
- Bundler:          `Vite`
- User Config:      `publicRuntimeConfig`, `meta`, `router`, `generate`, `css`, `components`, `buildModules`, `modules`, `build`, `vite`, `nitro`
- Runtime Modules:  `-`
- Build Modules:    `@pinia/nuxt@0.1.8`
------------------------------

code in my plugins is the same as yours. 

> import { defineNuxtPlugin } from '#app'
import { useRoute } from '#imports'
export default defineNuxtPlugin((nuxtApp) => {
  const route = useRoute()
  // this should now work as you wish
  console.log('route', JSON.stringify(route))
})

and the console result is 
>{"path":"/","params":{},"query":{},"hash":"","fullPath":"/","matched":[],"meta":{}}

Any solution about this? THX

**Reply by danielroe** [reply]
Would you raise an issue about this? 🙏

**Reply by foolsoft** [reply]
Same problem. Got 
`"path":"/","params":{},"query":{},"hash":"","fullPath":"/","matched":[],"meta":{}}` 
by any url 

`export default defineNuxtPlugin((app) => {
  const nuxtRoute = useRoute();
  console.log(JSON.stringify(nuxtRoute)); ....`

UPD: but as i found - inside functions that's works fine

export default defineNuxtPlugin(() => {
  const nuxtRoute = useRoute();
  
  return {
     provide: {
       path: () =>  { 
           console.log(nuxtRoute.params); //OK
       }
    }
  }
})
